Closes #157 | Create dataset loader for M3Exam

MJonibek opened the pull request:
Closes #157 

Note: zipfile library is used to unzip data, because password is needed. Could not find how pass password using dl_manager.download_and_extract() function.

Please name your PR after the issue it closes. You can use the following line: "Closes #ISSUE-NUMBER" where you replace the ISSUE-NUMBER with the one corresponding to your dataset.

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py`.
- [ ] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

sabilmakbar commented:
> zipfile library is used to unzip data, because password is needed. Could not find how pass password using dl_manager.download_and_extract() function.

In that case, would you like to add this lib into SEACrowd Package Reqs in a separate PR? Many thanks!

MJonibek commented:
> > zipfile library is used to unzip data, because password is needed. Could not find how pass password using dl_manager.download_and_extract() function.
> 
> In that case, would you like to add this lib into SEACrowd Package Reqs in a separate PR? Many thanks!

I think no need in this, since it is Python`s module (like os)

jamesjaya commented on seacrowd/sea_datasets/m3exam/m3exam.py:
```diff
@@ -0,0 +1,177 @@
+        for lang in _LANGUAGES
+    ]
+
+    DEFAULT_CONFIG_NAME = f"{_DATASETNAME}_source"
```
DEFAULT_CONFIG_NAME must be populated with one of the config names, maybe just put one of the _LANGUAGES as default

MJonibek replied:
Yes, sorry missed it

jamesjaya commented on seacrowd/sea_datasets/m3exam/m3exam.py:
```diff
@@ -0,0 +1,177 @@
+        if not os.path.exists(data_dir + ".zip"):
+            os.rename(data_dir, data_dir + ".zip")
+
+        with zipfile.ZipFile(data_dir + ".zip", "r") as zip_ref:
+            zip_ref.extractall(data_dir, pwd=_PASSWORD)  # unzipping with password
```
I got this error while trying to call `load_dataset`, could you check? Let me know if you can't reproduce the issue

```
>>> x = datasets.load_dataset('seacrowd/sea_datasets/m3exam/m3exam.py', name='m3exam_tha_source')
Downloading data: 100%|█████████████████████████████████████████████████████████████████████████| 2.38k/2.38k [00:00<00:00, 1.94MB/s]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jamesjaya/Library/Python/3.9/lib/python/site-packages/datasets/load.py", line 2152, in load_dataset
    builder_instance.download_and_prepare(
  File "/Users/jamesjaya/Library/Python/3.9/lib/python/site-packages/datasets/builder.py", line 948, in download_and_prepare
    self._download_and_prepare(
  File "/Users/jamesjaya/Library/Python/3.9/lib/python/site-packages/datasets/builder.py", line 1711, in _download_and_prepare
    super()._download_and_prepare(
  File "/Users/jamesjaya/Library/Python/3.9/lib/python/site-packages/datasets/builder.py", line 1021, in _download_and_prepare
    split_generators = self._split_generators(dl_manager, **split_generators_kwargs)
  File "/Users/jamesjaya/.cache/huggingface/modules/datasets_modules/datasets/m3exam/79e5824e0f329a471260f71ab881767300f3ca5a8d06f60c4efd6ff6fef1f5ec/m3exam.py", line 120, in _split_generators
    with zipfile.ZipFile(data_dir + ".zip", "r") as zip_ref:
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/zipfile.py", line 1257, in __init__
    self._RealGetContents()
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/lib/python3.9/zipfile.py", line 1324, in _RealGetContents
    raise BadZipFile("File is not a zip file")
zipfile.BadZipFile: File is not a zip file
```

MJonibek replied:
Ok, I will work on it this week

MJonibek replied:
@jamesjaya I found out the reason of this problem. While using dl_manager.download(urls) instead of downloading dataset it downloads google page.
![001](https://github.com/SEACrowd/seacrowd-datahub/assets/44943993/bacf2ce2-22e9-4fad-811d-18f1f9100671)

I will try do resolve this problem without using additional libraries. But it seems I will need to use gdown library (if I will use gdown library need I create another PR for adding it to requirements.txt?)

MJonibek replied:
@jamesjaya done changes, should work now

MJonibek commented:
Hi @holylovenia, done changes with meta feature

holylovenia commented:
Quick question, @MJonibek. The [M3Exam](https://github.com/DAMO-NLP-SG/M3Exam) dataset's examples show that some questions come with an image. Considering we're using the `qa` schema, are the SEA subsets completely in natural language only (without an image)?

MJonibek commented:
@holylovenia Yes, all questions are without any images. When you download a .zip file, it has two directories: multimodal-questions and text-questions. I use only a text-questions directory that consists only of text questions without any images (completely natural language).

MJonibek commented:
> Hmmm, could you please check the `multimodal-questions` to see whether our current languages have some data over there? It would be great to have a visual question answering subset if there are.

Yes, it has Javanese, Vietnamese and Thai

MJonibek commented:
If I need to add subsets with multimodal-questions, can you say which schema and task constant to use? Right now we have:

    IMAGE_CAPTIONING = "IC"
    SIGN_LANGUAGE_RECOGNITION = "SLR"
    STYLIZED_IMAGE_CAPTIONING = "SIC"
    VISUALLY_GROUNDED_REASONING = "VGR"

I think we need to create a new task IMAGE_QA. For schema, we can use the same qa schema, by adding the path to the image in meta features.

holylovenia commented:
> > Hmmm, could you please check the `multimodal-questions` to see whether our current languages have some data over there? It would be great to have a visual question answering subset if there are.
> 
> Yes, it has Javanese, Vietnamese and Thai

Could you please add them, @MJonibek? Previously, you might have to add `Tasks.VISUAL_QUESTION_ANSWERING` and a new schema (maybe we can name it `imqa` to retain the `im-` prefix? 😂)—a blend of `imtext` and `qa`. I'll see if I can do it quickly if you don't have the time though.

What do you think?

cc: @sabilmakbar @SamuelCahyawijaya @jamesjaya about the new task and schema.

MJonibek commented:
Yes, I can. I can do it today or tomorrow. We need to agree towards new schema features. Here draft version (just add image_paths to qa schema):

features = datasets.Features(
    {
        "id": datasets.Value("string"),
        "question_id": datasets.Value("string"),
        "document_id": datasets.Value("string"),
        "question": datasets.Value("string"),
        "type": datasets.Value("string"),
        "choices": datasets.Sequence(datasets.Value("string")),
        "context": datasets.Value("string"),
        "answer": datasets.Sequence(datasets.Value("string")),
        "image_paths": datasets.Sequence(datasets.Value("string")),

        "meta": {}
    }
)

holylovenia commented:
> Yes, I can. I can do it today or tomorrow. We need to agree towards new schema features. Here draft version (just add image_paths to qa schema):
> 
> features = datasets.Features( { "id": datasets.Value("string"), "question_id": datasets.Value("string"), "document_id": datasets.Value("string"), "question": datasets.Value("string"), "type": datasets.Value("string"), "choices": datasets.Sequence(datasets.Value("string")), "context": datasets.Value("string"), "answer": datasets.Sequence(datasets.Value("string")), "image_paths": datasets.Sequence(datasets.Value("string")), "meta": {} } )

This looks good to me. IMO, `"questions": datasets.Sequence(datasets.Value("string"))" would be better since it can serve documents with multiple questions; some VQA datasets are built this way, e.g., COCO (not a SEA dataset).

If @sabilmakbar @SamuelCahyawijaya @jamesjaya have some feedback, please let us know.

MJonibek commented:
Hi @holylovenia @jamesjaya 

I added question-answering schema. But I am not sure if is it good idea to divide source schema into 2 parts: one for text qa with name m3exam_{lang}_source (it has two splits valid and test), another for image qa with name m3exam_{lang}_imqa_source (it has only one split). 

I think, I can mix two source schemas into one with the name m3exam_{lang}_source where the split train will be for image qa, and splits valid and test for text qa.

holylovenia commented:
> another for image qa with name m3exam{lang}_imqa_source (it has only one split)
> 
> I think, I can mix two source schemas into one with the name m3exam_{lang}_source where the split train will be for image qa, and splits valid and test for text qa.

If all of the data in this dataset follows this rule, I agree to do it this way, @MJonibek.

MJonibek commented:
@holylovenia I merged two source schemas into one with the name m3exam_{lang}_source where the split train will be for image qa, and splits valid and test for text qa.

holylovenia reviewed (changes requested):
Hi @MJonibek, thanks for the changes and the implementation this dataset. A few suggestions:

1. Could you please remove the leading letter from the answer options in `*_seacrowd_qa` and `*_seacrowd_imqa`? For example:

From this
```
{
  'answer': ['D. Remo'],
  'choices': ['A. Dangdutan', 'B. Saman', 'C. Kecak', 'D. Remo'],
... }
```

To this
```
{
  'answer': ['Remo'],
  'choices': ['Dangdutan', 'Saman', 'Kecak', 'Remo'],
... }
```

2. Could you please fix the `image_paths` in both `source` and `seacrowd_imqa` schemas? The current `image_paths` have "png" twice and a `(` and `)`.

From this
```
{ ...
  'image_paths': ["~/.cache/huggingface/datasets/downloads/521b0e8f089704a1fb1c8040de9e9a289c6d7655b0c21ec5d3193f35f5729faa_extracted/data/multimodal-question/images-javanese/image-('0', "
                 "'png').png"],
... }
```

To this
```
{ ...
  'image_paths': ["~/.cache/huggingface/datasets/downloads/521b0e8f089704a1fb1c8040de9e9a289c6d7655b0c21ec5d3193f35f5729faa_extracted/data/multimodal-question/images-javanese/image-0.png"],
... }
```

MJonibek commented:
Hi @holylovenia, done changes, sorry about the image paths bug

holylovenia reviewed (approved):
Thank you for fixing them, @MJonibek! LGTM. Merging now.

holylovenia commented:
I noticed that some of the `m3exam_tha_seacrowd_qa`'s data instances have `[]` as their `answer` in the `test` split (e.g., those with ID `8, 20, 25, 26, 32, ..., 2080, 2100`). I didn't check the `train` and `validation`, so I don't know if this also applies to them.

Would it be okay for you to make a pull request for the fix? I would also understand if you don't have the bandwidth for this.

Sorry for noticing so late, @MJonibek. 🙏

MJonibek commented:
Hello @holylovenia , I found the problem. It is because of inconsistency in the dataset: for example in the sample with ID 8 answer is 2, however numeration in the answer options is in Thai and saved as  \u0e51, \u0e52, \u0e53, \u0e54. As a result, when dataloader is searching answer_text in the answer_options, it can not find the right answer. Any idea how to solve it? Right now it is:

`"answer": [".".join(answer.split(".")[1:]) for answer in json_obj["options"] if json_obj["answer_text"] == answer.split(".")[0]]`
